internal/pkg/addon/package.go

```diff
@@ -0,0 +1,302 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package addon
+
+import (
+	"archive/zip"
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/fs"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/aws/copilot-cli/internal/pkg/aws/s3"
+	"github.com/aws/copilot-cli/internal/pkg/template/artifactpath"
+	"gopkg.in/yaml.v3"
+)
+
+type uploader interface {
+	Upload(bucket, key string, data io.Reader) (string, error)
+}
+
+type transformInfo struct {
+	Property           []string
+	BucketNameProperty string
+	ObjectKeyProperty  string
```

Comment:
We need more explanation on these. It took me a while to understand property is a list of string but they are not in parallel with each other but a path chain instead. and `BucketNameProperty` `ObjectKeyProperty` are only for end node schema names.